Progress Report:
- DONE: Feb 14: can output a file written with comments but error in delimiters, every character delimited by ','
    Also, Python error about being unable to transcribe unicode for a certain string about 3 secs into loop in get_comments()

In [1]:
import json
import pandas as pd
import requests
import numpy as np
import pandas as pd
import csv
import datetime

In [2]:
# set parameters for request

subreddit = 'asianamerican'
limit = 100
timeframe = 'all' #hour, day, week, month, year, all
listing = 'top' # controversial, best, hot, new, random, rising, top

### get_reddit(): returns .json object

In [13]:
def get_reddit(subreddit,listing,limit,timeframe):
    try:
        base_url = f'https://www.reddit.com/r/{subreddit}/{listing}.json?limit={limit}&t={timeframe}'
        request = requests.get(base_url, headers = {'User-agent': 'yourbot'})
    except:
        print('An Error Occured')
    return request.json()

In [14]:
r = get_reddit(subreddit, listing, limit, timeframe)

In [15]:
def get_post_titles(r):
    '''
    Get a List of post titles
    '''
    posts = []
    for post in r['data']['children']:
        x = post['data']['title']
        posts.append(x)
    return posts

In [6]:
titles = get_post_titles(r)

In [7]:
print(len(titles))

100


In [8]:
print(titles[0:5])

['NBA All-Star Damian Lillard wears ""Stop Asian Hate" shirt while sitting out tonight\'s game', '“Fuck you! We will stop the Hate!” NBA star Baron Davis stands against anti-Asian racism.', 'Accurate', "My friend's mother was one of the victims of Anti-Asian shooting that recently occurred. The gofundme already lays out their situation but she was a single mother working hard to support her kids through school and the pandemic. Any contributions would mean the world", 'Megan Thee Stallion donates $50K to AAPI legal advocacy group after Atlanta spa shootings']


In [9]:
# print keys/elements of a Reddit object (in this case, submission (t3))
list_red_objs = list(search['data']['children'][0]['data'].keys())

# important keys: (1) id (Reddit thing ID), (2) title, (3) ups (upvotes), 
# (4) downs (downvotes), (5) score (not sure), (6) likes (not sure), (7) num_comments)

NameError: name 'search' is not defined

In [10]:
# What categories do I need from each post?
## Comments
## 

#print(r['data']['children'])
dict_ = pd.DataFrame.from_dict(r['data']['children'][0])
dict_[0:40]

kind         data
all_awardings                   t3           []
allow_live_comments             t3         True
approved_at_utc                 t3         None
approved_by                     t3         None
archived                        t3         True
author                          t3     Shiggy_O
author_flair_background_color   t3         None
author_flair_css_class          t3         None
author_flair_richtext           t3           []
author_flair_template_id        t3         None
author_flair_text               t3         None
author_flair_text_color         t3         None
author_flair_type               t3         text
author_fullname                 t3   t2_nu91ii0
author_is_blocked               t3        False
author_patreon_flair            t3        False
author_premium                  t3        False
awarders                        t3           []
banned_at_utc                   t3         None
banned_by                       t3         None
can_gild                        t3        False
can_mod_post                    t3        False
category                        t3         None
clicked                         t3        False
content_categories              t3         None
contest_mode                    t3        False
created                         t3  1.61681e+09
created_utc                     t3  1.61681e+09
discussion_type                 t3         None
distinguished                   t3         None
domain                          t3    i.redd.it
downs                           t3            0
edited                          t3        False
gilded                          t3            0
gildings                        t3           {}
hidden                          t3        False
hide_score                      t3        False
id                              t3       me5ef9
is_created_from_ads_ui          t3        False
is_crosspostable                t3        False

### get_search()
- gets the .json file of a Reddit search

In [16]:
# Corrected request url
#https://www.reddit.com/r/asianamerican/search/?q=affirmative%20action&restrict_sr=1&sort=top

def get_search(subreddit,query,limit,timeframe,sort,restrict_sr):
    try:
        base_url = f'https://www.reddit.com/r/{subreddit}/search.json?q={query}&restrict_sr={restrict_sr}&limit={limit}&t={timeframe}&sort={sort}'
        request = requests.get(base_url, headers = {'User-agent': 'yourbot'})
    except:
        print('An Error Occured')
    return request.json()

### Get top 10 post (submission) IDs from Search

In [17]:
def get_top_n_submission_info(r, n):
    '''
    Get a List of post titles
    r: .json object
    n: num of top posts
    # important keys: (1) id (Reddit thing ID), (2) title, (3) ups (upvotes), 
    # (4) downs (downvotes), (5) score (not sure), (6) likes (not sure), (7) num_comments)

    Returns: pandas Dataframe object
    '''
    # create 3 empty arrays of length n
    id_arr = np.empty(n, dtype="S10")
    up_arr = np.empty(n)
    down_arr = np.empty(n)
    score_arr = np.empty(n)
    likes_arr = np.empty(n)
    num_com_arr = np.empty(n)
    
    for i, post in enumerate(r['data']['children']):
        id_arr[i] = post['data']['id']
        up_arr[i] = post['data']['ups']
        down_arr[i] = post['data']['downs']
        score_arr[i] = post['data']['score']
        likes_arr[i] = post['data']['likes']
        num_com_arr[i] = post['data']['num_comments']

    # create Dataframe
    df = pd.DataFrame({'id': id_arr, 'upvotes': up_arr, 'downvotes': down_arr, 'score': score_arr, 'likes': likes_arr, 'num_comments': num_com_arr})


    return df

#### Driver file

In [18]:
# get_search() params
sr = 'asianamerican'
q = 'affirmative action'
limit = 100
timeframe = 'all'
sort = 'top'
restrict_sr = 1

# variable for array of titles
titles = None

search = get_search(subreddit=sr,query=q, limit=limit, timeframe=timeframe, sort=sort, restrict_sr=restrict_sr)
df = get_top_n_submission_info(search, limit)

In [19]:
df.head(50)

id  upvotes  downvotes  score  likes  num_comments
0   b'14m8mf4'    242.0        0.0  242.0    NaN         357.0
1    b'hqic5b'    228.0        0.0  228.0    NaN         114.0
2    b'9nk0do'    209.0        0.0  209.0    NaN          44.0
3    b'sejv5m'    149.0        0.0  149.0    NaN          85.0
4    b'h9tuc2'    137.0        0.0  137.0    NaN         108.0
5    b'sbrca3'    135.0        0.0  135.0    NaN         224.0
6    b'cwk0vc'    130.0        0.0  130.0    NaN         136.0
7    b'8ajkwc'    121.0        0.0  121.0    NaN          32.0
8    b'8usex7'    117.0        0.0  117.0    NaN          42.0
9    b'rjy8ut'    107.0        0.0  107.0    NaN          39.0
10   b'7481yg'     96.0        0.0   96.0    NaN         116.0
11   b'5btr2j'     92.0        0.0   92.0    NaN          23.0
12   b'87ru7q'     79.0        0.0   79.0    NaN          81.0
13   b'pvb846'     79.0        0.0   79.0    NaN          30.0
14   b'jny9p4'     72.0        0.0   72.0    NaN         139.0
15   b'3oymmf'     61.0        0.0   61.0    NaN          65.0
16   b'6rolbg'     57.0        0.0   57.0    NaN          42.0
17   b'8uonyt'     50.0        0.0   50.0    NaN          36.0
18   b'sz1qyo'     48.0        0.0   48.0    NaN          19.0
19   b'9meo0b'     47.0        0.0   47.0    NaN          65.0
20   b'8vt7kt'     46.0        0.0   46.0    NaN          30.0
21   b'8rlr3y'     45.0        0.0   45.0    NaN          42.0
22   b'j7ybkv'     38.0        0.0   38.0    NaN           3.0
23   b'9mzffm'     40.0        0.0   40.0    NaN          75.0
24   b'jm4x1b'     32.0        0.0   32.0    NaN          27.0
25   b'b9rf6b'     32.0        0.0   32.0    NaN           6.0
26   b'ti11r7'     32.0        0.0   32.0    NaN          66.0
27   b'3w7268'     30.0        0.0   30.0    NaN           6.0
28   b'6ruvgx'     29.0        0.0   29.0    NaN          51.0
29   b'967i7r'     28.0        0.0   28.0    NaN          58.0
30   b'9ozm49'     29.0        0.0   29.0    NaN         219.0
31   b'9wwgwt'     29.0        0.0   29.0    NaN           9.0
32   b'6t02ts'     29.0        0.0   29.0    NaN           4.0
33   b'6r0tss'     25.0        0.0   25.0    NaN          59.0
34   b'9z3ssz'     27.0        0.0   27.0    NaN          24.0
35   b'5cudq1'     29.0        0.0   29.0    NaN           6.0
36   b'3dmbpz'     24.0        0.0   24.0    NaN           8.0
37   b'sbvysg'     26.0        0.0   26.0    NaN           4.0
38   b'6rjw6f'     25.0        0.0   25.0    NaN          11.0
39   b'62q795'     23.0        0.0   23.0    NaN           1.0
40   b'9s7sz1'     24.0        0.0   24.0    NaN          14.0
41   b'3e4d2v'     22.0        0.0   22.0    NaN           2.0
42   b'36jedc'     24.0        0.0   24.0    NaN          60.0
43   b'4kq3uy'     21.0        0.0   21.0    NaN           9.0
44   b'5i6idf'     21.0        0.0   21.0    NaN          58.0
45   b'dezvi6'     21.0        0.0   21.0    NaN          25.0
46   b'9kgf7i'     20.0        0.0   20.0    NaN          40.0
47   b'21a4zz'     17.0        0.0   17.0    NaN          38.0
48   b'9oq200'     19.0        0.0   19.0    NaN           0.0
49   b'23wexw'     17.0        0.0   17.0    NaN          28.0

In [20]:
# get column of t3 object submission IDs
id_arr = df['id'].to_numpy()

# decode strings from byte 
def decode_arr(x):
    return x.decode('utf-8')

vfunc = np.vectorize(decode_arr)
id_arr_decoded = vfunc(id_arr)

print(id_arr_decoded)

['14m8mf4' 'hqic5b' '9nk0do' 'sejv5m' 'h9tuc2' 'sbrca3' 'cwk0vc' '8ajkwc'
 '8usex7' 'rjy8ut' '7481yg' '5btr2j' '87ru7q' 'pvb846' 'jny9p4' '3oymmf'
 '6rolbg' '8uonyt' 'sz1qyo' '9meo0b' '8vt7kt' '8rlr3y' 'j7ybkv' '9mzffm'
 'jm4x1b' 'b9rf6b' 'ti11r7' '3w7268' '6ruvgx' '967i7r' '9ozm49' '9wwgwt'
 '6t02ts' '6r0tss' '9z3ssz' '5cudq1' '3dmbpz' 'sbvysg' '6rjw6f' '62q795'
 '9s7sz1' '3e4d2v' '36jedc' '4kq3uy' '5i6idf' 'dezvi6' '9kgf7i' '21a4zz'
 '9oq200' '23wexw' '3euiwo' '6riosv' '31whah' '20ewl2' '1yujng' '9rd4tx'
 '23ouhl' 'dskb09' 'anfd61' 'hlnwtr' 'ldzz0f' '6ra6ey' '105fx43' '9oovpa'
 'jka8ii' '2n01k4' '1v63z3' '8a1j3p' 'h107c4' '2rdduz' '1juzmv' '9ap203'
 'drhgp6' '6sqhhh' 'sujd6e' '3rgzi3' '12e8uno' '3bkavj' 'tuqh5o' '43eu4h'
 'mauapi' '88rs2g' '266qsv' '3ija6r' 'aulovs' '14v43k' '4eb7qj' '2c8iu1'
 '3rkhqd' '4q20ib' '24hmx8' '4lqzqq' 'esbwk7' 'd6767e' '6ek03l' '5cw7uu'
 '3ed2ls' '1belws6' 'bgj9bw' '4ziaf0']


Use id_arr to get comments of these submission id's.

## Get comments

### Using Reddit API

In [23]:
'''
# Don't use below code - use PRAW instead
def get_comments(sr, post_id, context, limit, showedits, showmedia, showmore, showtitle, theme, threaded, truncate, sort):
    try:
        base_url = f'https://www.reddit.com/r/{sr}/comments/{post_id}?context={context}&limit={limit}&showedits={showedits}&showmore={showmore}&showmedia={showmedia}&showtitle={showtitle}&sort={sort}&theme={theme}&threaded={threaded}&truncate={truncate}'
        request = requests.get(base_url, headers = {'User_agent': 'yourbot'})
        return request.json()
    except:
        print('An Error Occured')
'''

In [21]:
'''
sr = 'asianamerican'
context = 5 # int: 0-8
limit = 100
showedits = False
showmore = False
showmedia = False
showtitle = False
sort = 'confidence' # one of (confidence, top, new, controversial, old, random, qa, live)
theme = 'default'
threaded = False
truncate = 25
post_id = 'sejv5m'
'''

In [24]:

'''comments = get_comments(sr=sr,
                        post_id=post_id,
                        context=context,
                        limit=limit,
                        showedits=showedits,
                        showmore=showmore,
                        showtitle=showtitle,
                        showmedia=showmedia,
                        sort=sort,
                        theme=theme,
                        threaded=threaded,
                        truncate=truncate)'''

An Error Occured


### Using PRAW Instead of Reddit API

In [3]:
import praw

# need to take secret and password off of Github in the future

reddit = praw.Reddit(
    client_id="iaKUGuo77BwJYHElyAo1Jg",
    client_secret="19bSH4KJpPJB2WFeYhUe--jk4IKeHQ",
    password='Ehdwn1928K!',
    user_agent="Comment Extraction u/dlehdwn",
    username='dlehdwn',
)

In [13]:
print(reddit.user.me())

dlehdwn


In [71]:
#enable read-only mode
#reddit.read_only = True

In [12]:
def get_comments(praw_obj, id_arr, filename):
    '''
    praw_obj: PRAW Reddit object
    id_arr: array of Reddit t3 IDs
    '''
    #open a file to write to
    with open(filename, 'w', newline='', encoding="utf-8") as f:

        # writer
        writer = csv.writer(f, delimiter=',', quoting=csv.QUOTE_MINIMAL)

        # for each submission/post
        for id in id_arr:
            sub = praw_obj.submission(id)
            
            # get post data
            time = datetime.datetime.fromtimestamp(sub.created)
            post_text = sub.title + ' ' + sub.selftext
            parent_id = None

            # submission author can be None
            if sub.author is not None:
                username = sub.author.name
                flair = sub.author_flair_text
                writer.writerow([id, parent_id, username, time, flair, post_text])

            else:
                #print('comment.author is None')
                #print(f'body: {body}\n')
                writer.writerow([id, parent_id, None, time, None, post_text])

            # get all comments
            sub.comments.replace_more(limit=None, threshold=1) #threshold - min num of comments needed to traverse into "More comments" object
            for comment in sub.comments.list():

                # get comment features
                comment_time = datetime.datetime.fromtimestamp(comment.created)
                comment_id = comment.id
                comment_parent_id = comment.parent_id
                body = comment.body

                # get username of comment author
                if comment.author is not None:
                    comment_username = comment.author.name
                    comment_flair = comment.author_flair_text
                    writer.writerow([comment_id, comment_parent_id, comment_username, comment_time, comment_flair, body])
                
                else:
                    #print('comment.author is None')
                    #print(f'body: {body}\n')
                    writer.writerow([comment_id, comment_parent_id, None, comment_time, None, body])

    f.close()

In [21]:
num_comments_collected = get_comments(reddit, id_arr_decoded, 'top_100_post_comments_user_time_text.txt') #takes a min to run

In [35]:
# num of actual total comments
print(df['num_comments'].sum())

# num of collected comments
print(num_comments_collected)

3709.0
3633


#### Num of comments collected
- Num of collected posts and comments in top 100 posts: 3633
- Num of actual total posts and comments in top 100 posts: 3709

### Example with one submission

In [4]:
submission = reddit.submission("14m8mf4")

In [5]:
import datetime

# get datetime of post
created = submission.created
time = datetime.datetime.fromtimestamp(created)
print(time)

2023-06-29 10:54:44


In [11]:
if submission.author is not None:
    username = submission.author.name
    flair = submission.author_flair_text
    time = datetime.datetime.fromtimestamp(submission.created)
    time_utc = datetime.datetime.fromtimestamp(submission.created_utc) # created and created_utc return the same time
    self_text = submission.selftext
    title = submission.title
    id = submission.id

print(id)
print(title)
print(username)
print(flair)
print(time)
print(time_utc)
print(self_text)

14m8mf4
[Megathread] Supreme Court Ruling on Affirmative Action
Tungsten_
None
2023-06-29 10:54:44
2023-06-29 10:54:44
This is a consolidated thread for users to discuss today's supreme court decision on affirmative action at Harvard and UNC. Please, even in disagreement, be civil and kind.

[NBC](https://www.nbcnews.com/politics/supreme-court/supreme-court-strikes-affirmative-action-programs-harvard-unc-rcna66770)

[CNN](https://www.cnn.com/politics/live-news/supreme-court-decisions/index.html)

[NYT](https://www.nytimes.com/live/2023/06/29/us/affirmative-action-supreme-court)

[WaPo](https://www.washingtonpost.com/education/2023/06/29/supreme-court-student-loan-forgiveness-affirmative-action/)

[Supreme Court Opinion](https://www.supremecourt.gov/opinions/22pdf/20-1199_hgdj.pdf)


In [6]:
# print top level comments
for top_level_comment in submission.comments:
    print(top_level_comment.body)

Thanks to everyone who engaged in insightful and respectful discourse about the news. 

This thread is now locked for comments.
I would prefer using a process that takes into account poverty instead. 

The first generation of my family that came to America was *painfully* poor and everyone showed up with neither money nor education. They worked in kitchens and laundromats. Notice a lot of people in bigger reddit boards talking shit about the "Chinese billionaire" boogeyman (fearmongering like this also erases the less visible Asian races who came to America as refugees and reduces all Asians to a monolithic "rich Asian stereotype") and how this will only help them. The Chinese people I know were not coming to America with bags of cash.
u/Tungsten_, Thanks for creating a section just to discuss this. When I read the news I immediately went searching for a forum where folks might have civil discourse on this topic.

Just had a few comments/questions:

1. Has anyone come across seemingly 

In [10]:
submission.comments.replace_more(limit=None)
for comment in submission.comments.list()[:10]:
    #print(comment.body)
    #print(datetime.datetime.fromtimestamp(comment.created))
    print(comment.parent_id)
    

t3_14m8mf4
t3_14m8mf4
t3_14m8mf4
t3_14m8mf4
t3_14m8mf4
t3_14m8mf4
t3_14m8mf4
t3_14m8mf4
t3_14m8mf4
t3_14m8mf4
